compiler/src/dotty/tools/dotc/transform/SpecializeFunctions.scala

```diff
@@ -0,0 +1,231 @@
+        }
+
+        cpy.Template(tmpl)(parents = specializedParents, body = body0)
+      }
+      case _ => tree
+    }
+
+  /** Dispatch to specialized `apply`s in user code */
+  override def transformApply(tree: Apply)(implicit ctx: Context, info: TransformerInfo) = {
+    import ast.Trees._
+    tree match {
+      case Apply(select @ Select(id, nme.apply), arg :: Nil) =>
+        val params = List(arg.tpe, tree.tpe)
+        val specializedApply = nme.apply.specializedFor(params, params.map(_.typeSymbol.name), Nil, Nil)
+        val hasOverridenSpecializedApply = id.tpe.decls.iterator.exists { sym =>
+          sym.is(Flags.Override) && (sym.name eq specializedApply)
+        }
+
+        if (hasOverridenSpecializedApply) tpd.Apply(tpd.Select(id, specializedApply), arg :: Nil)
+        else tree
+      case _ => tree
+    }
+  }
+
+  private def specializedName(name: Name, args: List[Type], ret: Type)(implicit ctx: Context): Name = {
+    val typeParams = args :+ ret
+    name.specializedFor(typeParams, typeParams.map(_.typeSymbol.name), Nil, Nil)
+  }
+
+  // Extractors ----------------------------------------------------------------
+  private object ShouldTransformDenot {
+    def unapply(cref: ClassDenotation)(implicit ctx: Context): Option[Seq[SpecializationTarget]] =
+      if (!cref.classParents.map(_.symbol).exists(defn.isPlainFunctionClass)) None
+      else Some(getSpecTargets(cref.typeRef))
+  }
+
+  private object ShouldTransformTree {
+    def unapply(tree: Template)(implicit ctx: Context): Option[Seq[(Tree, SpecializationTarget)]] = {
+      val treeToTargets = tree.parents
+        .map(t => (t, getSpecTargets(t.tpe)))
+        .filter(_._2.nonEmpty)
+        .map { case (t, xs) => (t, xs.head) }
+
+      if (treeToTargets.isEmpty) None else Some(treeToTargets)
+    }
+  }
+
+  private case class SpecializationTarget(target: Symbol,
+                                          params: List[Type],
+                                          ret: Type,
+                                          original: Symbol)
+
+  /** Gets all valid specialization targets on `tpe`, allowing multiple
+   *  implementations of FunctionX traits
+   */
+  private def getSpecTargets(tpe: Type)(implicit ctx: Context): List[SpecializationTarget] = {
+    val functionParents =
+      tpe.classSymbols.iterator
+        .flatMap(_.baseClasses)
+        .filter(defn.isPlainFunctionClass)
+
+    val tpeCls = tpe.widenDealias
+    functionParents.map { sym =>
+      val typeParams = tpeCls.baseArgTypes(sym)
+      val args = typeParams.init
+      val ret = typeParams.last
+
+      val interfaceName =
+        (functionName ++ args.length)
+        .specializedFor(typeParams, typeParams.map(_.typeSymbol.name), Nil, Nil)
+
+      val interface = ctx.getClassIfDefined(functionPkg ++ interfaceName)
+
+      if (interface.exists) Some {
+        SpecializationTarget(interface, args, ret, sym)
+      }
+      else None
+    }
+    .flatten.toList
```

Comment:
why toList. Do you care what iterable kind you get?
Conversion costs time and memory